operators/src/processing/new_expression/mod.rs

```diff
@@ -0,0 +1,749 @@
+        );
+
+        let sources = self.sources.initialize(context).await?;
+
+        let spatial_reference = sources.a.result_descriptor().spatial_reference;
+
+        for other_spatial_refenence in sources
+            .iter()
+            .skip(1)
+            .map(|source| source.result_descriptor().spatial_reference)
+        {
+            ensure!(
+                spatial_reference == other_spatial_refenence,
+                crate::error::InvalidSpatialReference {
+                    expected: spatial_reference,
+                    found: other_spatial_refenence,
+                }
+            );
+        }
+
+        let result_descriptor = RasterResultDescriptor {
+            data_type: self.params.output_type,
+            spatial_reference,
+            measurement: self
+                .params
+                .output_measurement
+                .as_ref()
+                .map_or(Measurement::Unitless, Measurement::clone),
+            no_data_value: Some(self.params.output_no_data_value), // TODO: is it possible to have none?
+        };
+
+        let initialized_operator = InitializedExpression {
+            result_descriptor,
+            sources,
+            expression,
+            map_no_data: self.params.map_no_data,
+        };
+
+        Ok(initialized_operator.boxed())
+    }
+}
+
+pub struct InitializedExpression {
+    result_descriptor: RasterResultDescriptor,
+    sources: ExpressionInitializedSources,
+    expression: ExpressionAst,
+    map_no_data: bool,
+}
+
+pub struct ExpressionInitializedSources {
+    a: Box<dyn InitializedRasterOperator>,
+    b: Option<Box<dyn InitializedRasterOperator>>,
+    c: Option<Box<dyn InitializedRasterOperator>>,
+}
+
+impl ExpressionInitializedSources {
+    fn iter(&self) -> impl Iterator<Item = &Box<dyn InitializedRasterOperator>> {
+        let mut sources = vec![&self.a];
+
+        if let Some(o) = self.b.as_ref() {
+            sources.push(o);
+        }
+
+        if let Some(o) = self.c.as_ref() {
+            sources.push(o);
+        }
+
+        sources.into_iter()
+    }
+}
+
+impl InitializedRasterOperator for InitializedExpression {
+    fn query_processor(&self) -> Result<TypedRasterQueryProcessor> {
+        let output_type = self.result_descriptor().data_type;
+        // TODO: allow processing expression without NO DATA
+        let output_no_data_value = self.result_descriptor().no_data_value.unwrap_or_default();
+
+        let expression = LinkedExpression::new(&self.expression)?;
+
+        let query_processors: Vec<TypedRasterQueryProcessor> = self
+            .sources
+            .iter()
+            .map(InitializedRasterOperator::query_processor)
+            .collect::<Result<_>>()?;
+
+        Ok(match query_processors.len() {
+            1 => {
+                let [a] = <[_; 1]>::try_from(query_processors).expect("len previously checked");
+                call_on_generic_raster_processor!(a, p_a => {
```

Comment:
why is this not turned into_f64?

Reply:
Since we don't have such an explosion of symbols here, I thought we could spare the extra conversion.

Reply:
can't that lead to different results if I have an expression on one raster and it operates on the original type and I add a second input and now it operates on float or do I misunderstand it?

Reply:
46856e5